src/Quasar/Advanced/QuasarAF.purs

```diff
@@ -19,87 +19,295 @@ module Quasar.Advanced.QuasarAF
   , module Quasar.QuasarF
   , module Quasar.Error
   , module Quasar.Types
+  , module Quasar.Advanced.Types
   ) where
 
 import Prelude
 
 import Data.Argonaut (JArray)
-import Data.Either (Either)
+import Data.Foldable (class Foldable, foldMap)
 import Data.Functor.Coproduct (Coproduct, left, right)
 import Data.Maybe (Maybe)
+import Data.Path.Pathy as Pt
 
-import Quasar.Advanced.Auth.Provider as Auth
 import Quasar.Data (QData, JSONMode(..))
 import Quasar.Data.Json.Extended (EJson, resultsAsEJson)
-import Quasar.Error (QError(..), lowerQError, printQError)
+import Quasar.Error (type (:~>), QError, QResponse, lowerQError, printQError)
 import Quasar.FS (Resource)
 import Quasar.Mount (MountConfig)
 import Quasar.QuasarF (QuasarF(..))
 import Quasar.Query.OutputMeta (OutputMeta)
 import Quasar.ServerInfo (ServerInfo)
 import Quasar.Types (AnyPath, FilePath, DirPath, Pagination, Vars, SQL)
+import Quasar.Advanced.Types as Qa
 
-type QuasarAFP = Coproduct QuasarF QuasarAF
 
 data QuasarAF a
-  = AuthProviders (Either QError (Array Auth.Provider) → a)
+  = GroupInfo (Pt.AbsFile Pt.Sandboxed) (Qa.GroupInfo :~> a)
+  | CreateGroup (Pt.AbsFile Pt.Sandboxed) (Unit :~> a)
+  | ModifyGroup (Pt.AbsFile Pt.Sandboxed) Qa.GroupPatch (Unit :~> a)
+  | DeleteGroup (Pt.AbsFile Pt.Sandboxed) (Unit :~> a)
+  | PermissionList Boolean ((Array Qa.Permission) :~> a)
+  | PermissionInfo Qa.PermissionId (Qa.Permission :~> a)
+  | PermissionChildren Qa.PermissionId Boolean ((Array Qa.Permission) :~> a)
+  | SharePermission Qa.ShareRequest ((Array Qa.Permission) :~> a)
+  | DeletePermission Qa.PermissionId (Unit :~> a)
+  | TokenList ((Array Qa.Token) :~> a)
+  | TokenInfo Qa.TokenId (Qa.Token :~> a)
+  | CreateToken (Maybe Qa.TokenName) (Array Qa.Action) (Qa.Token :~> a)
+  | DeleteToken Qa.TokenId (Unit :~> a)
+  | AuthProviders ((Array Qa.Provider) :~> a)
```

Comment:
You probably noticed there was a problem with how tightly the type operators bind, that'll be fixed in 0.9...